shapely/coordinates.py

```diff
@@ -1,30 +1,76 @@
+    geometry,
+    transformation,
+    include_z: Union[bool, int] = False,
+    single_vec: bool = True,
+    rebuild: bool = False,
+):
+    """Returns a copy of a geometry array with a function applied to its coordinates.
+
+    With the default of ``rebuild=False``, ``include_z=False``,
+    all returned geometries will be two-dimensional;
+    the third dimension will be discarded, if present.
+    When specifying ``rebuild=False``, ``include_z=True``,
+    the returned geometries preserve the dimensionality of the
+    respective input geometries.
+    When specifying ``rebuild=True`` the output geometries
+    will get rebuilt from the transformation output
 
     Parameters
     ----------
     geometry : Geometry or array_like
     transformation : function
-        A function that transforms a (N, 2) or (N, 3) ndarray of float64 to
+        A function that transforms coordinates, in ``single_vec=True`` mode:
+        the function transforms a (N, 2) or (N, 3) ndarray of float64 to
         another (N, 2) or (N, 3) ndarray of float64.
-    include_z : bool, default False
-        If True, include the third dimension in the coordinates array
-        that is passed to the ``transformation`` function. If a
-        geometry has no third dimension, the z-coordinates passed to the
-        function will be NaN.
+        in ``single_vec=False`` mode: the function transform 2 or 3 N ndarray
+        of float64 to another 2 or 3 ndarray of float64.
+        With the default ``rebuild=False``, the transform function must also
+        return N sized array(s).
+    include_z : bool or int, default False
+        If True | 1, include the third dimension in the coordinates array
+            that is passed to the ``transformation`` function. If a
+            geometry has no third dimension, the z-coordinates passed to the
+            function will be NaN.
+        If False | 0, The third-dimension of the geometries (if exist),
+            would not be included in the input to transform function.
+        If -1 Include the third dimension in the coordinates array that is passed to the
+            ``transformation`` function only if the geoemtries are three-dimensional.
+    single_vec: bool, default True
+        How to pass coordinates to the transform function:
+        if True, pass them as a single argument
+            (2D array in case of multiple coordinates)
+        Otherwise, Pass them as separate x, y(, z) arguments
+            (separate 1D arrays in case of multiple coordinates)
+    rebuild: bool, default False
+        If False, output geometries will preserve the coordinate number and
+            dimnensionality as the input geometries (but in include_z=False,
+            two-dimension geometries are always returned).
+        Otherwise, output geometries will get rebuilt from the transformation
+            output, so they might have different number of coordinate or
+            a different coordiante-dimensionality.
```

Comment:
General note on formatting: this way of indenting paragraphs doesn't work for restructuredtext (it would be interpreted as "definition list" (https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#lists-and-quote-like-blocks), but that would render incorrectly)